src/gen_rules.ml

```diff
@@ -355,16 +355,17 @@ let gen ~contexts
         Path.Map.find (Install_rules.packages sctx) path))
   in
   Build_system.set_rule_generators
-    (function
-      | Install ctx ->
-        Option.map (String.Map.find sctxs ctx) ~f:(fun sctx ->
-          (fun ~dir _ ->
-             Install_rules.gen_rules sctx ~dir:(Path.as_in_build_dir_exn dir);
-             Build_system.These String.Set.empty))
-      | Context ctx ->
-        String.Map.find generators ctx);
-
-  String.Map.iter map ~f:(fun (module M : Gen) ->
-    Build_system.handle_add_rule_effects (fun () ->
-      Odoc.init M.sctx));
+    ~init:(fun () ->
+      String.Map.iter map ~f:(fun (module M : Gen) ->
+        Build_system.handle_add_rule_effects (fun () ->
+          Odoc.init M.sctx)))
```

Comment:
Why can't we call `Odoc.init` anymore? The `~init` argument could use a comment in build_system.mli

Reply:
The point of this change is to make `handle_add_rule_effects` only happen in build_system.ml, not here.

There was a commit that does it (219ec2d5d0ae4ddf934f8f3a8f618ef9152539d7) that's logically part of this PR, but I moved it into another PR by mistake. I moved it back here now.

Reply:
Added a comment to `init`: [init] can generate rules in any directory, so it's always called.

And to directly answer the question of why we can't just call `Odoc.init` directly, it's because it produces (via implicit output) and the only way to give rules to the build system now is by calling `set_rule_generators`.